### ISSUE TITLE
__WEBPACK_IMPORTED_MODULE_3_jquery__ is not a function

### ISSUE THREAD

**Issue by TristanVeyet** [dougludlow/ng2-bs3-modal#147]
Can replicate using your own angular-cli project in few steps:
```
git clone https://github.com/dougludlow/ng2-bs3-modal-demo-angular-cli.git
ng build
ng serve
```
Project compiles and works.
As in my project i need bootstrap (for responsiveness) and jQuery, i added typings this way:
```
npm install @types/bootstrap -S
npm install @types/jquery -S
```
I use jQuery in modal-demo.component.ts:
```
import { Component, ViewChild, ViewEncapsulation, OnInit } from '@angular/core';
...
import * as jquery from 'jquery';

@Component({
    selector: 'modal-demo-component',
   ...
})
export class ModalDemoComponent implements OnInit {
    ...
    constructor(private router: Router) { }

    ngOnInit() {
      jquery('h1').text('Hello Angular2');
    }
```

And then:
```
ng build
ng serve
```
Project compiles but don't works.
There are errors in the console:
`TypeError: __WEBPACK_IMPORTED_MODULE_3_jquery__ is not a function`

If i try to suppress jquery reference in angular-cli.json:
```
"styles": [
        "../node_modules/bootstrap/dist/css/bootstrap.css",
        "styles.css"
      ],
      "scripts": [
        "../node_modules/bootstrap/js/modal.js"
      ],
```
I've got another error:
`ReferenceError: jQuery is not defined`  as referenced in the issue #132 

what's going wrong? Is there a way to fix it?  :roll_eyes:

**Comment by dougludlow** [comment]
Bootstrap requires that jQuery is global, a limitation of Bootstrap. I'm guessing you'll need to use it globally throughout your project as webpack doesn't want to include it twice.

Instead of:

```
import * as jquery from 'jquery';
```

Do this:

```
declare var jQuery: any;
```

**Comment by TristanVeyet** [comment]
It's a pity, i used `import * as jquery from 'jquery';` for typings and intellisense.
I won't be able to use it anymore.

**Comment by dougludlow** [comment]
@TristanVeyet give this a try: 

```
declare vat jQuery: jQuery;
```

You can give it a type instead of just `any`.

**Comment by TristanVeyet** [comment]
Thanks for your hint but give it a type require to import Jquery and it's precisely what's cause me a problem because reference is loaded twice
- First by the script loading in the angular-cli.json (that's required by the bootstrap library)
- and then by the import * as jQuery from 'jquery';

I found this link that is interresting to understand he problem:
https://weblog.west-wind.com/posts/2016/Sep/12/External-JavaScript-dependencies-in-Typescript-and-Angular-2#jQueryandPlugins

But it doesn't solved my problem. For the moment, i use `declare var $: any;` and don't use intellisense

**Comment by dougludlow** [comment]
@TristanVeyet you don't need to import a library in order to get the typing for it. Simply including the typing file and referencing the type is enough.

**Comment by TristanVeyet** [comment]
@dougludlow if i don't import the jQuery library, declare a variable as jquery like you said: `declare var jQuery: jQuery;` throw an error.
But when i declare no variable, $ and jQuery are available (i guess that importing jquery script in angular-cli.json make them global) so its solves my problem.
Thanks for your help and congrats for your job.

**Comment by kanchansingh1123** [comment]
Is there any solution around it ? facing same issue.

**Comment by thegust** [comment]
Any solutions?

**Comment by dougludlow** [comment]
@thegust try this:

1. Create a `jquery.ts` in the same location as your `main.ts` and paste the following in it:
   ```ts
   import * as $ from 'jquery';
   window['jQuery'] = window['$'] = $;
   ```
2. Insert the following at the top of your `main.ts`:
   ```ts
   import './jquery';
   import 'bootstrap';
   ```
3. Remove any other script references to jquery or bootstrap from your html or angular cli config.

**Comment by C-V-K-D** [comment]
Hey !

I got a similar issue but with **_moment_** instead of **_jQuery_** :
`Error: Uncaught (in promise): TypeError: __WEBPACK_IMPORTED_MODULE_3_moment__ is not a function`

I was trying to use **_ion2-calendar_** (https://www.npmjs.com/package/ion2-calendar),
and found the exact same issue in the owner's github issues **with the solution** :
 ==> https://github.com/HsuanXyz/ion2-calendar/issues/83

In short, the new version of **_ion2-calendar_** is not yet compatible with the new version of **_moment_**, and both were released the same day (yesterday), so when I installed both of them I had this issue.

Hope this will help !

**Comment by dougludlow** [comment]
@C-V-K-D - that appears to be an error with moment and isn't related to ng2-bs3-modal at all. ng2-bs3-modal doesn't use moment.

**Comment by C-V-K-D** [comment]
@dougludlow Oh okay... Thought it might be related, my bad !

**Comment by amorel** [comment]
I don't know why but this work for me:

```
import $ from 'jquery'
declare var $: $
```

**Comment by srinivasan-g** [comment]
Thanks  amorel 
this working

**Comment by amorel** [comment]
another way is to install Jquery Types:
`npm install @types/jquery`

and use this in the app.module to use Jquery globaly:
`import * as $ from 'jquery';`

**Comment by FanFanFantastic** [comment]
Holy Sh* , you saved my life. Why does adding it to  the app.module.ts works????!

another way is to install Jquery Types:
npm install @types/jquery

and use this in the app.module to use Jquery globaly:
import * as $ from 'jquery';

**Comment by danderoyal** [comment]
THANK YOU @amorel !!!
I used this solution, and i added `declare var $: any;` inside the component i needed to use jquery for. 
I've tried about a billion of other solutions but this was the only thing working for me.

**Comment by sumit-sharma7** [comment]
Thanks a lot @amorel ! You're a life saver.
declare var $: any; worked like a charm!
